commit b7e024a838452a85870256d8f1ab946dc8f931cd
Author: Stefan Liebler <stli@linux.ibm.com>
Date:   Tue Dec 18 13:57:05 2018 +0100

    S390: Refactor memcmp ifunc handling.
    
    This patch moves all ifunc variants for memcmp
    to sysdeps/s390/memcmp-z900.S. The configure-check/preprocessor logic
    in sysdeps/s390/ifunc-memcmp.h decides if ifunc is needed at all
    and which ifunc variants should be available.
    E.g. if the compiler/assembler already supports z196 by default,
    the older ifunc variants are not included.
    If we only need the newest ifunc variant,
    then we can skip ifunc at all.
    
    Therefore the ifunc-resolvers and __libc_ifunc_impl_list are adjusted
    in order to handle only the available ifunc variants.
    
    ChangeLog:
    
            * sysdeps/s390/ifunc-memcmp.h: New File.
            * sysdeps/s390/memcmp.S: Move to ...
            * sysdeps/s390/memcmp-z900.S ... here.
            Move implementations from memcmp-s390x.s to here.
            * sysdeps/s390/multiarch/memcmp-s390x.S: Delete File.
            * sysdeps/s390/multiarch/Makefile (sysdep_routines):
            Remove memcmp variants.
            * sysdeps/s390/Makefile (sysdep_routines):
            Add memcmp variants.
            * sysdeps/s390/multiarch/ifunc-impl-list.c
            (__libc_ifunc_impl_list): Adjust ifunc variants for
            memcmp.
            * sysdeps/s390/multiarch/memcmp.c: Move ifunc resolver
            to ...
            * sysdeps/s390/memcmp.c: ... here.
            Adjust ifunc variants for memcmp.

diff --git a/sysdeps/s390/Makefile b/sysdeps/s390/Makefile
index e40e5bd982e54d89..c59cbdc25aad408a 100644
--- a/sysdeps/s390/Makefile
+++ b/sysdeps/s390/Makefile
@@ -31,5 +31,6 @@ sysdeps-gconv-modules = ../sysdeps/s390/gconv-modules
 endif
 
 ifeq ($(subdir),string)
-sysdep_routines += bzero memset memset-z900
+sysdep_routines += bzero memset memset-z900 \
+		   memcmp memcmp-z900
 endif
diff --git a/sysdeps/s390/ifunc-memcmp.h b/sysdeps/s390/ifunc-memcmp.h
new file mode 100644
index 0000000000000000..536ac455d177c027
--- /dev/null
+++ b/sysdeps/s390/ifunc-memcmp.h
@@ -0,0 +1,59 @@
+/* memcmp variant information on S/390 version.
+   Copyright (C) 2018 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#if defined USE_MULTIARCH && IS_IN (libc)	\
+  && ! defined HAVE_S390_MIN_Z196_ZARCH_ASM_SUPPORT
+# define HAVE_MEMCMP_IFUNC	1
+#else
+# define HAVE_MEMCMP_IFUNC	0
+#endif
+
+#if defined HAVE_S390_MIN_Z196_ZARCH_ASM_SUPPORT
+# define MEMCMP_DEFAULT		MEMCMP_Z196
+# define HAVE_MEMCMP_Z900_G5	0
+# define HAVE_MEMCMP_Z10	0
+# define HAVE_MEMCMP_Z196	1
+#elif defined HAVE_S390_MIN_Z10_ZARCH_ASM_SUPPORT
+# define MEMCMP_DEFAULT		MEMCMP_Z10
+# define HAVE_MEMCMP_Z900_G5	0
+# define HAVE_MEMCMP_Z10	1
+# define HAVE_MEMCMP_Z196	HAVE_MEMCMP_IFUNC
+#else
+# define MEMCMP_DEFAULT		MEMCMP_Z900_G5
+# define HAVE_MEMCMP_Z900_G5	1
+# define HAVE_MEMCMP_Z10	HAVE_MEMCMP_IFUNC
+# define HAVE_MEMCMP_Z196	HAVE_MEMCMP_IFUNC
+#endif
+
+#if HAVE_MEMCMP_Z900_G5
+# define MEMCMP_Z900_G5		__memcmp_default
+#else
+# define MEMCMP_Z900_G5		NULL
+#endif
+
+#if HAVE_MEMCMP_Z10
+# define MEMCMP_Z10		__memcmp_z10
+#else
+# define MEMCMP_Z10		NULL
+#endif
+
+#if HAVE_MEMCMP_Z196
+# define MEMCMP_Z196		__memcmp_z196
+#else
+# define MEMCMP_Z196		NULL
+#endif
diff --git a/sysdeps/s390/multiarch/memcmp-s390x.S b/sysdeps/s390/memcmp-z900.S
similarity index 54%
rename from sysdeps/s390/multiarch/memcmp-s390x.S
rename to sysdeps/s390/memcmp-z900.S
index 6321737acec821ec..c3b3677ba78264ee 100644
--- a/sysdeps/s390/multiarch/memcmp-s390x.S
+++ b/sysdeps/s390/memcmp-z900.S
@@ -1,4 +1,4 @@
-/* CPU specific memcmp implementations.  31/64 bit S/390 version.
+/* memcmp - compare two memory blocks.  31/64 bit S/390 version.
    Copyright (C) 2012-2018 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
 
@@ -17,8 +17,9 @@
    <http://www.gnu.org/licenses/>.  */
 
 
-#include "sysdep.h"
+#include <sysdep.h>
 #include "asm-syntax.h"
+#include <ifunc-memcmp.h>
 
 /* INPUT PARAMETERS
      %r2 = address of first memory area
@@ -27,46 +28,67 @@
 
        .text
 
-#if IS_IN (libc)
-
-ENTRY(__memcmp_z196)
-	.machine "z196"
-	.machinemode "zarch_nohighgprs"
-# if !defined __s390x__
-	llgfr	%r4,%r4
-# endif /* !defined __s390x__  */
-	ltgr    %r4,%r4
-	je      .L_Z196_4
-	aghi    %r4,-1
+#if HAVE_MEMCMP_Z900_G5
+# if defined __s390x__
+#  define LTGR	ltgr
+#  define AGHI	aghi
+#  define BRCTG	brctg
+# else
+#  define LTGR	ltr
+#  define AGHI	ahi
+#  define BRCTG	brct
+# endif /* ! defined __s390x__  */
+ENTRY(MEMCMP_Z900_G5)
+# if defined __s390x__
+	.machine "z900"
+# else
+	.machine "g5"
+	basr    %r5,0
+.L_Z900_G5_16:
+#  define Z900_G5_EX_D .L_Z900_G5_15-.L_Z900_G5_16
+# endif /* ! defined __s390x__  */
+	LTGR    %r4,%r4
+	je      .L_Z900_G5_4
+	AGHI    %r4,-1
+# if defined __s390x__
 	srlg    %r1,%r4,8
-	ltgr    %r1,%r1
-	jne     .L_Z196_2
-.L_Z196_3:
-	exrl    %r4,.L_Z196_14
-.L_Z196_4:
+	larl    %r5,.L_Z900_G5_15
+#  define Z900_G5_EX_D 0
+# else
+	lr	%r1,%r4
+	srl	%r1,8
+# endif /* ! defined __s390x__  */
+	LTGR    %r1,%r1
+	jne     .L_Z900_G5_12
+.L_Z900_G5_3:
+	ex      %r4,Z900_G5_EX_D(%r5)
+.L_Z900_G5_4:
 	ipm     %r2
+# if defined __s390x__
 	sllg    %r2,%r2,34
 	srag    %r2,%r2,62
+# else
+	sll     %r2,2
+	sra     %r2,30
+# endif /* ! defined __s390x__  */
 	br      %r14
-.L_Z196_17:
+.L_Z900_G5_12:
+	clc     0(256,%r3),0(%r2)
+	jne     .L_Z900_G5_4
 	la      %r3,256(%r3)
 	la      %r2,256(%r2)
-	aghi    %r1,-1
-	je      .L_Z196_3
-.L_Z196_2:
-	pfd     1,512(%r3)
-	pfd     1,512(%r2)
-	clc     0(256,%r3),0(%r2)
-	je      .L_Z196_17
-	ipm     %r2
-	sllg    %r2,%r2,34
-	srag    %r2,%r2,62
-	br      %r14
-.L_Z196_14:
+	BRCTG   %r1,.L_Z900_G5_12
+	j       .L_Z900_G5_3
+.L_Z900_G5_15:
 	clc     0(1,%r3),0(%r2)
-END(__memcmp_z196)
+END(MEMCMP_Z900_G5)
+# undef LTGR
+# undef AGHI
+# undef BRCTG
+#endif /* HAVE_MEMCMP_Z900_G5  */
 
-ENTRY(__memcmp_z10)
+#if HAVE_MEMCMP_Z10
+ENTRY(MEMCMP_Z10)
 	.machine "z10"
 	.machinemode "zarch_nohighgprs"
 # if !defined __s390x__
@@ -95,18 +117,57 @@ ENTRY(__memcmp_z10)
 	j       .L_Z10_3
 .L_Z10_15:
 	clc     0(1,%r3),0(%r2)
-END(__memcmp_z10)
+END(MEMCMP_Z10)
+#endif /* HAVE_MEMCMP_Z10  */
 
-#endif /* IS_IN (libc) */
+#if HAVE_MEMCMP_Z196
+ENTRY(MEMCMP_Z196)
+	.machine "z196"
+	.machinemode "zarch_nohighgprs"
+# if !defined __s390x__
+	llgfr	%r4,%r4
+# endif /* !defined __s390x__  */
+	ltgr    %r4,%r4
+	je      .L_Z196_4
+	aghi    %r4,-1
+	srlg    %r1,%r4,8
+	ltgr    %r1,%r1
+	jne     .L_Z196_2
+.L_Z196_3:
+	exrl    %r4,.L_Z196_14
+.L_Z196_4:
+	ipm     %r2
+	sllg    %r2,%r2,34
+	srag    %r2,%r2,62
+	br      %r14
+.L_Z196_17:
+	la      %r3,256(%r3)
+	la      %r2,256(%r2)
+	aghi    %r1,-1
+	je      .L_Z196_3
+.L_Z196_2:
+	pfd     1,512(%r3)
+	pfd     1,512(%r2)
+	clc     0(256,%r3),0(%r2)
+	je      .L_Z196_17
+	ipm     %r2
+	sllg    %r2,%r2,34
+	srag    %r2,%r2,62
+	br      %r14
+.L_Z196_14:
+	clc     0(1,%r3),0(%r2)
+END(MEMCMP_Z196)
+#endif /* HAVE_MEMCMP_Z196  */
 
-#include "../memcmp.S"
+#if ! HAVE_MEMCMP_IFUNC
+/* If we don't use ifunc, define an alias for memcmp here.
+   Otherwise see sysdeps/s390/memcmp.c.  */
+strong_alias (MEMCMP_DEFAULT, memcmp)
+weak_alias (memcmp, bcmp)
+#endif
 
-#if !IS_IN (libc)
-.globl   memcmp
-.set     memcmp,__memcmp_default
-.weak    bcmp
-.set	 bcmp,__memcmp_default
-#elif defined SHARED && IS_IN (libc)
-.globl   __GI_memcmp
-.set     __GI_memcmp,__memcmp_default
+#if defined SHARED && IS_IN (libc)
+/* Defines the internal symbols.
+   Compare to libc_hidden_builtin_def (memcmp) in string/memcmp.c.  */
+strong_alias (MEMCMP_DEFAULT, __GI_memcmp)
 #endif
diff --git a/sysdeps/s390/memcmp.S b/sysdeps/s390/memcmp.S
deleted file mode 100644
index 751293a99e34f530..0000000000000000
--- a/sysdeps/s390/memcmp.S
+++ /dev/null
@@ -1,96 +0,0 @@
-/* memcmp - compare two memory blocks.  31/64 bit S/390 version.
-   Copyright (C) 2012-2018 Free Software Foundation, Inc.
-   This file is part of the GNU C Library.
-
-   The GNU C Library is free software; you can redistribute it and/or
-   modify it under the terms of the GNU Lesser General Public
-   License as published by the Free Software Foundation; either
-   version 2.1 of the License, or (at your option) any later version.
-
-   The GNU C Library is distributed in the hope that it will be useful,
-   but WITHOUT ANY WARRANTY; without even the implied warranty of
-   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-   Lesser General Public License for more details.
-
-   You should have received a copy of the GNU Lesser General Public
-   License along with the GNU C Library; if not, see
-   <http://www.gnu.org/licenses/>.  */
-
-
-#include <sysdep.h>
-#include "asm-syntax.h"
-
-/* INPUT PARAMETERS
-     %r2 = address of first memory area
-     %r3 = address of second memory area
-     %r4 = number of bytes to compare.  */
-
-       .text
-#if defined __s390x__
-# define LTGR	ltgr
-# define AGHI	aghi
-# define BRCTG	brctg
-#else
-# define LTGR	ltr
-# define AGHI	ahi
-# define BRCTG	brct
-#endif /* ! defined __s390x__  */
-
-#ifdef USE_MULTIARCH
-ENTRY(__memcmp_default)
-#else
-ENTRY(memcmp)
-#endif
-#if defined __s390x__
-	.machine "z900"
-#else
-	.machine "g5"
-	basr    %r5,0
-.L_Z900_G5_16:
-# define Z900_G5_EX_D .L_Z900_G5_15-.L_Z900_G5_16
-#endif /* ! defined __s390x__  */
-	LTGR    %r4,%r4
-	je      .L_Z900_G5_4
-	AGHI    %r4,-1
-#if defined __s390x__
-	srlg    %r1,%r4,8
-	larl    %r5,.L_Z900_G5_15
-# define Z900_G5_EX_D 0
-#else
-	lr	%r1,%r4
-	srl	%r1,8
-#endif /* ! defined __s390x__  */
-	LTGR    %r1,%r1
-	jne     .L_Z900_G5_12
-.L_Z900_G5_3:
-	ex      %r4,Z900_G5_EX_D(%r5)
-.L_Z900_G5_4:
-	ipm     %r2
-#if defined __s390x__
-	sllg    %r2,%r2,34
-	srag    %r2,%r2,62
-#else
-	sll     %r2,2
-	sra     %r2,30
-#endif /* ! defined __s390x__  */
-	br      %r14
-.L_Z900_G5_12:
-	clc     0(256,%r3),0(%r2)
-	jne     .L_Z900_G5_4
-	la      %r3,256(%r3)
-	la      %r2,256(%r2)
-	BRCTG   %r1,.L_Z900_G5_12
-	j       .L_Z900_G5_3
-.L_Z900_G5_15:
-	clc     0(1,%r3),0(%r2)
-#ifdef USE_MULTIARCH
-END(__memcmp_default)
-#else
-END(memcmp)
-libc_hidden_builtin_def (memcmp)
-weak_alias (memcmp, bcmp)
-#endif
-
-#undef LTGR
-#undef AGHI
-#undef BRCTG
diff --git a/sysdeps/s390/multiarch/memcmp.c b/sysdeps/s390/memcmp.c
similarity index 60%
rename from sysdeps/s390/multiarch/memcmp.c
rename to sysdeps/s390/memcmp.c
index 1e6f31806e172d7d..952ff6af7364fd92 100644
--- a/sysdeps/s390/multiarch/memcmp.c
+++ b/sysdeps/s390/memcmp.c
@@ -16,12 +16,34 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if IS_IN (libc)
+#include <ifunc-memcmp.h>
+#if HAVE_MEMCMP_IFUNC
 # define memcmp __redirect_memcmp
 # include <string.h>
 # undef memcmp
 # include <ifunc-resolve.h>
 
-s390_libc_ifunc (__redirect_memcmp, __memcmp, memcmp)
+# if HAVE_MEMCMP_Z900_G5
+extern __typeof (__redirect_memcmp) MEMCMP_Z900_G5 attribute_hidden;
+# endif
+
+# if HAVE_MEMCMP_Z10
+extern __typeof (__redirect_memcmp) MEMCMP_Z10 attribute_hidden;
+# endif
+
+# if HAVE_MEMCMP_Z196
+extern __typeof (__redirect_memcmp) MEMCMP_Z196 attribute_hidden;
+# endif
+
+s390_libc_ifunc_expr (__redirect_memcmp, memcmp,
+		      ({
+			s390_libc_ifunc_init ();
+			(HAVE_MEMCMP_Z196 && S390_IS_Z196 (stfle_bits))
+			  ? MEMCMP_Z196
+			  : (HAVE_MEMCMP_Z10 && S390_IS_Z10 (stfle_bits))
+			  ? MEMCMP_Z10
+			  : MEMCMP_DEFAULT;
+		      })
+		      )
 weak_alias (memcmp, bcmp);
 #endif
diff --git a/sysdeps/s390/multiarch/Makefile b/sysdeps/s390/multiarch/Makefile
index 53dd8654d73677db..c893ebc5659fd4ae 100644
--- a/sysdeps/s390/multiarch/Makefile
+++ b/sysdeps/s390/multiarch/Makefile
@@ -19,8 +19,7 @@ sysdep_routines += strlen strlen-vx strlen-c \
 		   rawmemchr rawmemchr-vx rawmemchr-c \
 		   memccpy memccpy-vx memccpy-c \
 		   memrchr memrchr-vx memrchr-c \
-		   mempcpy \
-		   memcmp memcmp-s390x
+		   mempcpy
 endif
 
 ifeq ($(subdir),wcsmbs)
diff --git a/sysdeps/s390/multiarch/ifunc-impl-list.c b/sysdeps/s390/multiarch/ifunc-impl-list.c
index 253f36045b57cc3c..2e57d01abc21522e 100644
--- a/sysdeps/s390/multiarch/ifunc-impl-list.c
+++ b/sysdeps/s390/multiarch/ifunc-impl-list.c
@@ -22,6 +22,7 @@
 #include <ifunc-impl-list.h>
 #include <ifunc-resolve.h>
 #include <ifunc-memset.h>
+#include <ifunc-memcmp.h>
 
 /* Maximum number of IFUNC implementations.  */
 #define MAX_IFUNC	3
@@ -78,12 +79,21 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
 	      )
 #endif /* HAVE_MEMSET_IFUNC */
 
+#if HAVE_MEMCMP_IFUNC
   IFUNC_IMPL (i, name, memcmp,
+# if HAVE_MEMCMP_Z196
 	      IFUNC_IMPL_ADD (array, i, memcmp,
-			      S390_IS_Z196 (stfle_bits), __memcmp_z196)
+			      S390_IS_Z196 (stfle_bits), MEMCMP_Z196)
+# endif
+# if HAVE_MEMCMP_Z10
 	      IFUNC_IMPL_ADD (array, i, memcmp,
-			      S390_IS_Z10 (stfle_bits), __memcmp_z10)
-	      IFUNC_IMPL_ADD (array, i, memcmp, 1, __memcmp_default))
+			      S390_IS_Z10 (stfle_bits), MEMCMP_Z10)
+# endif
+# if HAVE_MEMCMP_Z900_G5
+	      IFUNC_IMPL_ADD (array, i, memcmp, 1, MEMCMP_Z900_G5)
+# endif
+	      )
+#endif /* HAVE_MEMCMP_IFUNC */
 
 #ifdef SHARED
 
